### PR TITLE
Make Options::Parser serializable

### PR DESCRIPTION
**Pull request by wthrowe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Options/ParseOptions.hpp [hunk 1]
```diff
@@ -647,6 +653,25 @@ std::string Parser<OptionList, Group>::help() const noexcept {
...
+  static_assert(std::is_same_v<Group, NoSuchType>,
+                "Inner parsers should be recreated by the root parser, not "
+                "serialized.");
+  // We reconstruct most of the state when deserializing, rather than
+  // trying to package it.
+  p | help_text_;
+  p | input_source_;
+  if (p.isUnpacking() and not input_source_.empty()) {
+    auto received_source = std::move(input_source_);
+    input_source_.clear();
```
Comment (fmahebert) [inline src/Options/ParseOptions.hpp:670]: For my understanding, why are you clearing the `input_source_`? doesn't this mean the next attempt to checkpoint-restart will fail to restore the parser? And why first moving to a temporary `received_source` and then moving each element from there to the parser?

**Review by fmahebert — COMMENTED** [review]
For now a question after a quick look. This looks is pretty clean and will definitely be better than me trying to hack the Main chare to rebuild the parser. Thanks!

File: src/Options/ParseOptions.hpp (hunk 1 above)
Comment (wthrowe) [reply src/Options/ParseOptions.hpp:670]: Under normal operation, `input_source_` is populated as files are parsed.  During unpacking we run the entire parsing sequence again, which will, as a (side?) effect, repopulate `input_source_`.

Edit: This is why the tests run the serialization twice.

Comment (fmahebert) [reply src/Options/ParseOptions.hpp:670]: Thanks for clarifying! I'd suggest adding a short comment like `// reparsing the input sources will repopulate input_source_, so clear it first` or similar, above this line.

File: tests/Unit/Options/Test_Options.cpp [hunk 2]
```diff
@@ -1292,6 +1293,108 @@ SPECTRE_TEST_CASE("Unit.Options.overlay.subgroup_context", "[Unit][Options]") {
...
+    f(parser);
+    const auto parser2 = serialize_and_deserialize(parser);
+    f(parser2);
+    const auto parser3 = serialize_and_deserialize(parser2);
+    f(parser3);
+  };
+
+  {
+    // Groups
+    Options::Parser<tmpl::list<InnerGroupedTag, OuterGroupedTag, Simple>> opts(
```
Comment (fmahebert) [inline tests/Unit/Options/Test_Options.cpp]: [consistency nitpick] In the other test cases you call this `parser`

**Review by fmahebert — COMMENTED** [review]
After a closer look, just two minor optional comments. If you make changes, please squash directly.

**Review by fmahebert — COMMENTED** [review]
The fixup looks good, you can squash

**Comment by fmahebert** [comment]
Edit: this was post was wrong, nevermind.

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Pull request merged**